File: groupAnagrams.py
'''
   The approach is similar to isAnagram.py
'''

# Approach- Categorized by sorted strings 
import collections   
def groupAnagrams(strs):  # TC O(N k log k) // SC O(N k)  N = length of strs
	                                                    # k = max length of a string in strs
	ans=collections.defaultdict(list)  # values as list in the dictionary 
	for s in strs:                         # TC O(N)
		ans[tuple(sorted(s))].append(s)    # TC O(k log k)

	return ans.values()

# Approach- Categorize by Count 

def groupAnagrams1(strs):                # TC O(N k) // SC O(N k)
	ans = collections.defaultdict(list)
	for s in strs:
		count = [0] * 26
		for c in s:
			count[ord(c) - ord('a')] += 1
		ans[tuple(count)].append(s)
	return ans.values()

print(groupAnagrams(["eat","tea","tan","ate","nat","bat"]))
print(groupAnagrams1(["eat","tea","tan","ate","nat","bat"]))

File: groupAnagrams.java
import java.util.*;
// Approach 1- Categorize by sorted string 
class Solution { // TC O(n k log k) SC O(N*K)
    public List<List<String>> groupAnagrams(String[] strs) {
        if (strs.length == 0) return new ArrayList<>();
        Map<String, List> ans = new HashMap<String, List>();
        for (String s : strs) {
            char[] ca = s.toCharArray();
            Arrays.sort(ca);
            String key = String.valueOf(ca);
            if (!ans.containsKey(key)) 
                ans.put(key, new ArrayList<>());
            ans.get(key).add(s);
        }
        return new ArrayList(ans.values());
    }
}

// Approach 2- Categorize by count 
class Solution1{ // TC O(n*k) SC O(n*k)
    public List<List<String>> groupAnagrams(String[] strs) {
        if (strs.length == 0) return new ArrayList<>();
        Map<String, List> ans = new HashMap<String, List>();
        int[] count = new int[26];
        for (String s : strs) {
            Arrays.fill(count, 0);
            for (char c : s.toCharArray()) 
                count[c - 'a']++;

            StringBuilder sb = new StringBuilder("");
            for (int i = 0; i < 26; i++) {
                sb.append('#');
                sb.append(count[i]);
            }
            String key = sb.toString();
            if (!ans.containsKey(key)) 
                ans.put(key, new ArrayList<>());
            ans.get(key).add(s);
        }
        return new ArrayList(ans.values());
    }
}

class groupAnagrams{
    public static void main(String[] args) {
        String[] strs={"eat","tea","tan","ate","nat","bat"};
        Solution s= new Solution();
        System.out.println(s.groupAnagrams(strs));
        Solution1 s1= new Solution1();
        System.out.println(s1.groupAnagrams(strs));
    }
}